### PR TITLE
Plugin Changes to SONOS

### PR DESCRIPTION
**Pull request by sisamiwe**
- __init__.py
	- general:
		- implement f-Strings and remove .format()
		- harmonize log levels and log messages
		- sort and struct import of python modules
		- put all imports of exclusive python modules in try - except

	- class Speaker
		- add method favorite_radio_stations
		- add method sonos_favorites
		- add method favorite_radio_stations
		- add new methods in _zone_topology_event
		- remove hard-coded "Sonos" to plugin short-name

	- class Sonos(SmartPlugin)
		- bump version
		- __init__
			- adapt code to new plugin template
			- put getting plugin parameters in try-except
			- move handling of fixed speaker ips to own method
			- move checks and init of tts into own method
			- move reading and log of used Soco version to own method
			- move setting of Soco log-level to own method
			- add item_list parameter
		- run
			- add initial speaker discovery in run method - add scheduler for speaker discovery just if there are no fixed IPs given - parse_item - add valid item for plugin to item_list
		- add new method _check_webservice_port
		- add new method _check_webservice_ip
		- add new method _check_local_webservice_path_snippet
			- create path if given and not available
		- add new method _check_local_webservice_path
		- add new method _get_free_diskspace
		- add new method _init_webservice
		- add new method _init_tts
		- add new method _parse_speaker_ips
		- add new method get_soco_version
		- add new method _set_soco_logger _resolve_uid
			- change to do rekursive look-up of item with speaker uid up to 4 levels higher in item tree
		- add method _get_zone_name_from_uid
		- add property sonos_speaker

- plugin.yaml
	- bump to new version
	- adapt sh_minversion
	- adapt py_minversion
	- Clean formatting
	- add default to local_webservice_path
	- set value limits for webservice_port
	- add parameter webif_pagelength
	- enhance attribute description
	- add neu attributes
	- add items structs

- requirements.txt
	- remove packets moved to core requirements

- utils.py
	- formatting

- .webif\__init__.py
	- support webif data
	- index
		- add global_pagelength
		- change/enhance return content
	- get_data_html
		- add item values for updating web interface
	- add method discover to allow discovery by button on WebIF

- .webif\templates\index.html
	- implement webif according to current template

- .examples
    - moved to items structs in plugin.yaml

### CODE REVIEW

**Comment by sisamiwe** [comment]
@aschwith 
First PR as discussed is available

File: sonos/__init__.py [hunk 1]
```diff
@@ -2788,119 +2740,283 @@ def _handle_dpt3(self, item, caller=None, source=None, dest=None):
...
                     volume = self._resolve_child_command_int(item, 'tts_volume', -1)
                     fade_in = self._resolve_child_command_bool(item, 'tts_fade_in')
-                    sonos_speaker[uid].play_tts(item(), language, self._local_webservice_path, self._webservice_url,
-                                                volume, self._snippet_duration_offset, fade_in)
-                if command == 'play_snippet':
+                    sonos_speaker[uid].play_tts(item(), language, self._local_webservice_path, self._webservice_url, volume, self._snippet_duration_offset, fade_in)
+                elif command == 'play_snippet' and item() != "":
                     if item() == "":
-                        self.logger.error("No item value when executing play_snippet command")
+                        self.logger.error("No item value when executing 'play_snippet' command")
```
Comment (aschwith) [inline sonos/__init__.py:3012]: This will never be executed. Please doublecheck.

**Review by aschwith — COMMENTED** [review]
Please correct minor finding. The rest is ok.

File: sonos/__init__.py (hunk 1 above)
Comment (sisamiwe) [reply sonos/__init__.py:3012]: Hi, 
should I do the correction within this PR using an additional commit?

Comment (aschwith) [reply sonos/__init__.py:3012]: Yes please

Comment (sisamiwe) [reply sonos/__init__.py:3012]: Commit with correction is up

Comment (aschwith) [reply sonos/__init__.py:3012]: Great contribution.

**Pull request merged**